### PR TITLE
Change opaque type syntax from `existential type` to type alias `impl Trait`

### PR DESCRIPTION
**Pull request by varkor**
This implements a new feature gate `type_alias_impl_trait` (this is slightly different from the originally proposed feature name, but matches what has been used in discussion since), deprecating the old `existential_types` feature.

The syntax for opaque types has been changed. In addition, the "existential" terminology has been replaced with "opaque", as per previous discussion and the RFC.

This makes partial progress towards implementing https://github.com/rust-lang/rust/issues/63063.

r? @Centril

### CODE REVIEW

**Comment by alexreg** [comment]
👍 on the rename. What remains after this?

File: src/test/ui/type-alias-impl-trait.rs [hunk 1]
```diff
@@ -3,7 +3,7 @@
 #![allow(dead_code)]
```
Comment (Centril) [inline src/test/ui/type-alias-impl-trait.rs]: Duplicate file?

**Review by Centril — COMMENTED** [review]
r=me with review comments addressed

**Comment by Centril** [comment]
> This implements a new feature gate `type_alias_impl_trait` (this is slightly different from the originally proposed feature name, but matches what has been used in discussion since), deprecating the old `existential_types` feature.

After merging, can you please change the tracking issue description + the F-* label name?

File: src/test/ui/issues/issue-60662.stdout [hunk 2]
```diff
@@ -10,5 +10,5 @@ extern crate std;
 trait Animal { }
 
 fn main() {
-              pub existential type ServeFut : Animal;
+              pub type ServeFut= impl  : Animal;
```
[the remark this answers is not in the document]
Comment (alexreg) [reply src/test/ui/issues/issue-60662.stdout]: Not sure why GH messed up indentation here. And I think your suggested insertion is one column too far to the left, by accident. :-)

File: src/test/rustdoc-ui/coverage/traits.rs [hunk 3]
```diff
@@ -29,9 +29,10 @@ impl ThisTrait for SomeStruct {
 /// but what about those aliases? i hear they're pretty exotic
 pub trait MyAlias = ThisTrait + Send + Sync;
 
-// FIXME(58624): once rustdoc can process existential types, we need to make sure they're counted
-// /// woah, getting all existential in here
-// pub existential type ThisExists: ThisTrait;
+// FIXME(58624): once rustdoc can process opaque `impl Trait` types,
+// we need to make sure they're counted
+// /// woah, getting all abstract in here
```
[the remark this answers is not in the document]
Comment (varkor) [reply src/test/rustdoc-ui/coverage/traits.rs]: Doesn't really make sense, but I suppose the original didn't either.

[the remark this answers is not in the document]
Comment (Centril) [reply src/test/rustdoc-ui/coverage/traits.rs]: Yeah that joke is all opaque to me.

File: src/libsyntax/parse/parser.rs [hunk 4]
```diff
@@ -6808,40 +6802,29 @@ impl<'a> Parser<'a> {
...
-    ) -> PResult<'a, (Ident, AliasKind, ast::Generics)> {
+    /// Parses a type alias or opaque type.
+    fn parse_type_alias(&mut self) -> PResult<'a, (Ident, AliasKind, ast::Generics)> {
         let ident = self.parse_ident()?;
         let mut tps = self.parse_generics()?;
         tps.where_clause = self.parse_where_clause()?;
-        let alias = if existential {
-            self.expect(&token::Colon)?;
+        self.expect(&token::Eq)?;
+        let alias = if self.check_keyword(kw::Impl) {
```
[the remark this answers is not in the document]
Comment (varkor) [reply src/libsyntax/parse/parser.rs]: Yeah, this was simpler to implement immediately, as I thought the full implementation was going to require more work, and I wanted to get this first step done soon.

File: src/libsyntax/parse/parser.rs [hunk 5]
```diff
@@ -6808,40 +6802,29 @@ impl<'a> Parser<'a> {
         Ok(self.mk_item(lo.to(prev_span), invalid, ItemKind::ForeignMod(m), visibility, attrs))
     }
 
-    /// Parses `type Foo = Bar;`
-    /// or
-    /// `existential type Foo: Bar;`
-    /// or
-    /// `return `None``
+    /// Parses `type Foo = Bar;` or returns `None`
```
[the remark this answers is not in the document]
Comment (varkor) [reply src/libsyntax/parse/parser.rs]: (The sentence continues on the next line.)

File: src/librustc/hir/check_attr.rs [hunk 6]
```diff
@@ -27,7 +27,7 @@ pub(crate) enum Target {
     ForeignMod,
     GlobalAsm,
     Ty,
-    Existential,
+    OpaqueTy,
```
[the remark this answers is not in the document]
Comment (varkor) [reply src/librustc/hir/check_attr.rs]: Syntactically they are the same, but semantically they do something different. Perhaps I can experiment with merging them in the follow-up.

File: src/libsyntax/parse/parser.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Centril) [reply src/libsyntax/parse/parser.rs]: Yeah that's fair enough =)

File: src/libsyntax/parse/parser.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Centril) [reply src/libsyntax/parse/parser.rs]: Oh woops!

File: src/librustc/hir/lowering.rs [hunk 7]
```diff
@@ -1730,7 +1730,7 @@ impl<'a> LoweringContext<'a> {
         // Not tracking it makes lints in rustc and clippy very fragile, as
         // frequently opened issues show.
         let exist_ty_span = self.mark_span_with_reason(
```
[the remark this answers is not in the document]
Comment (varkor) [reply src/librustc/hir/lowering.rs]: Ah, you're right. I did search for "exist", but got so many irrelevant results I didn't find all the relevant ones. `exist_ty` is a better search though.

File: src/librustc/hir/check_attr.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Centril) [reply src/librustc/hir/check_attr.rs]: @varkor Can you file a follow up issue for that?... and let's resolve thereafter.

File: src/librustc/hir/mod.rs [hunk 8]
```diff
@@ -2421,17 +2421,17 @@ pub enum ItemKind {
     GlobalAsm(P<GlobalAsm>),
     /// A type alias, e.g., `type Foo = Bar<u8>`
     Ty(P<Ty>, Generics),
-    /// An existential type definition, e.g., `existential type Foo: Bar;`
-    Existential(ExistTy),
+    /// An opaque `impl Trait` type alias, e.g., `type Foo = impl Bar;`
```
[the remark this answers is not in the document]
Comment (varkor) [reply src/librustc/hir/mod.rs]: This is inconsistent with the rest of the `enum`.

[the remark this answers is not in the document]
Comment (Centril) [reply src/librustc/hir/mod.rs]: Alright sure.

File: src/librustc/infer/opaque_types/mod.rs [hunk 9]
```diff
@@ -40,7 +40,7 @@ pub struct OpaqueTypeDecl<'tcx> {
     /// for example:
     ///
     /// ```
-    /// existential type Foo;
+    /// type Foo = impl Baz;
```
[the remark this answers is not in the document]
Comment (varkor) [reply src/librustc/infer/opaque_types/mod.rs]: Good catch — I didn't even spot that `abstract type` was used throughout the compiler. All fixed now.

File: src/librustc/infer/opaque_types/mod.rs [hunk 10]
```diff
@@ -842,12 +842,12 @@ impl TypeFolder<'tcx> for ReverseMapper<'tcx> {
                 self.tcx.sess
                     .struct_span_err(
                         self.span,
-                        "non-defining existential type use in defining scope"
+                        "non-defining opaque type use in defining scope"
```
[the remark this answers is not in the document]
Comment (varkor) [reply src/librustc/infer/opaque_types/mod.rs]: Yeah, we definitely need to follow up with some diagnostics improvements :+1:

File: src/librustc_typeck/outlives/implicit_infer.rs [hunk 11]
```diff
@@ -296,7 +296,7 @@ pub fn check_explicit_predicates<'tcx>(
         // struct MyStruct<'x, X> { field: Box<dyn Trait<'x, X>> }
         // ```
         //
-        // The `where Self: 'a` predicate refers to the *existential, hidden type*
+        // The `where Self: 'a` predicate refers to the *opaque, hidden type*
```
[the remark this answers is not in the document]
Comment (varkor) [reply src/librustc_typeck/outlives/implicit_infer.rs]: Whoops, yes.

File: src/librustc/infer/opaque_types/mod.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Centril) [reply src/librustc/infer/opaque_types/mod.rs]: Can you file an issue for that + cc the usual suspects?

File: src/librustc/hir/check_attr.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (varkor) [reply src/librustc/hir/check_attr.rs]: Submitted https://github.com/rust-lang/rust/issues/63204.

File: src/librustc/infer/opaque_types/mod.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (varkor) [reply src/librustc/infer/opaque_types/mod.rs]: Submitted https://github.com/rust-lang/rust/issues/63205.

File: src/librustc/hir/print.rs [hunk 12]
```diff
@@ -565,16 +565,12 @@ impl<'a> State<'a> {
                 self.s.word(";");
                 self.end(); // end the outer ibox
             }
-            hir::ItemKind::Existential(ref exist) => {
-                self.head(visibility_qualified(&item.vis, "existential type"));
+            hir::ItemKind::OpaqueTy(ref opaque_ty) => {
```
Comment (Centril) [inline src/librustc/hir/print.rs]: Nit: consider refactoring out common parts of `::Ty` and `::OpaqueTy` using a higher order function + lambda for differences in the middle.

**Comment by Centril** [comment]
Changes look good.
r=me with the nit addressed. :)

**Comment by Centril** [comment]
Needs `--bless --compare-mode=nll` (so annoying...).

**Comment by varkor** [comment]
We need to do something about that. Though presumably non-NLL mode is going away before too long, so it might not be with changing now.

**Pull request merged**